pkg/storage/chunk/chunk.go

```diff
@@ -89,18 +88,20 @@ func NewChunk(userID string, fp model.Fingerprint, metric labels.Labels, c prom_
 // Post-checksums, externals keys become the same across DynamoDB, Memcache
 // and S3.  Numbers become hex encoded.  Keys look like:
 // `<user id>/<fingerprint>:<start time>:<end time>:<checksum>`.
+//
+// v12+, fingerprint is now a prefix to support better read and write request parallelization:
+// `<user>/<fprint>/<start>:<end>:<checksum>`
 func ParseExternalKey(userID, externalKey string) (Chunk, error) {
-	if !strings.Contains(externalKey, "/") {
+	if !strings.Contains(externalKey, "/") { // pre-checksum
 		return parseLegacyChunkID(userID, externalKey)
+	} else if strings.Count(externalKey, "/") == 2 { // v12+
```

Comment:
This a very hot path. Do we have a benchmark for this function ? If so we can we have a look at the price of the change. 

I have some idea of improvement but before I wonder how much the string.Count cost.

Reply:
Good call! I've fixed an existing benchmark and added two more for the old parsing conditional logic:

```
$ go test -bench=Key -benchmem
goos: darwin
goarch: amd64
pkg: github.com/grafana/loki/pkg/storage/chunk
cpu: VirtualApple @ 2.50GHz
BenchmarkParseNewerExternalKey-8         3107478               412.6 ns/op             0 B/op          0 allocs/op
BenchmarkParseNewExternalKey-8           2605483               392.4 ns/op             0 B/op          0 allocs/op
BenchmarkParseLegacyExternalKey-8        2919925               414.9 ns/op            48 B/op          1 allocs/op
BenchmarkParseOldLegacyExternalKey-8     2884390               412.2 ns/op            48 B/op          1 allocs/op
BenchmarkParseOldNewExternalKey-8        3363462               360.0 ns/op             0 B/op          0 allocs/op
PASS
ok      github.com/grafana/loki/pkg/storage/chunk       13.586s
```

Reply:
> Can you bench compare the root function too

I've added additional benchmarks for the root parsing functions!

Reply:
Thank you ! If you got some time, there's definitively improvement we can make once this is merged on that root  `ParseExternalKey`. I'm thinking we should try to use only string.Index and reslice as we go to avoid too many string.Index, I think we get it to work with a max of 2 string.Index, instead of string.Index and string.Count (which does a for loop).

We'll talk about this offline, now that we have a benchmark we can compare it, this is not urgent, but something fun you can take on.